### DISCUSSION TITLE
Unclear licence implications on produced binaries

### DISCUSSION THREAD

**Discussion by mirabilos** [open-watcom/open-watcom-v2#382] [General]
Hi,

as I currently read the licence file, binaries produced by your compiler do, as they include the startup files and excerpts from the C library, fall under the same licence as the compiler itself, with its relatively onerous (if I have a DOS program of 200K or so) requirements and the well-known Freeness issues.

Is there not any kind of runtime exception similar to other compilers’ which states that no restrictions are imposed on compiler/toolchain/linker output even despite the runtime library/support code being included?

**Comment by ArmstrongJ** [comment]
The Sybase Open Watcom Public License is what it is...  It has plenty of legal issues surrounding it, which is why it is not accepted by quite a few organizations.

Sadly, the only people who can change it are now [SAP, previously Sybase but acquired](https://en.wikipedia.org/wiki/Sybase) as they hold copyright over much of the code.  They would need to declare that the Sybase Open Watcom Public License version 2 is, preferably, a simple, existing license like BSD, MIT, etc.  Otherwise, the license just can't be changed.

There is _nobody_ from SAP/Sybase associated with this project at all at this point.  Feel free to contact them, though.

This bug is effectively a duplicate of #271

**Comment by mirabilos** [comment]
> This bug is effectively a duplicate of #271 

No, it’s not. Please let me explain the difference:

#271 is about people wishing to package Open Watcom for use in a Free ecosystem, and not liking its licence (I’m not going into more here, as it’s off-topic).

This one is about the following situation: I have a piece of software for DOS whose last build has been done over a decade ago by someone else, and I now wish to fix minor bugs in it. For that, I must be able to compile it (it’s supposed to use bcc (Borland C+ maybe?) 3.0 or perhaps 3.1, but I only have Turbo C 2.01 from the museum, so I’m looking for alternative compilers) and ship the result.

If the licence imposes on the binaries produced by this compiler, then I cannot legally ship the result (in this case because the program itself is under GPLv1, which forbids other restrictions from being added), making Open Watcom fail at the one foremost purpose it has (namely compiling programs).

**Comment by ArmstrongJ** [comment]
I only labeled it as a duplicate of #271 because it involves problems with the license terms. I understand you're looking for a runtime exception, but that's just one of many issues.

Again, nobody here can change the license. It would need to be updated by SAP/Sybase, a company that just isn't involved in this project any longer. Feel free to try, but nobody here has the legal rights to change _all_ of the runtime library's license terms.

**Comment by mirabilos** [comment]
>I only labeled it as a duplicate of #271 because it involves problems
>with the license terms. I understand you're looking for a runtime
>exception, but that's just one of many issues.

OK. Well, I was asking whether there was one, because I could not find
one and was surprised because I had expected there to be one.

>Again, nobody here can change the license. It would need to be updated
>by SAP/Sybase, a company that just isn't involved in this project any
>longer. Feel free to try, but nobody here has the legal rights to
>change _all_ of the runtime library's license terms.

OK, thanks for the clarification. So there is indeed none?

**Comment by ArmstrongJ** [comment]
> OK, thanks for the clarification. So there is indeed none?

Not to my knowledge.  I will point out that the omission of a runtime exception is effectively a mistake, not some sort of odd trick to get access to your source.  I realize that doesn't make strict licensing possible, but it makes most people feel a bit better about using Open Watcom

**Comment by mirabilos** [comment]
>> OK, thanks for the clarification. So there is indeed none?
>
>Not to my knowledge.  I will point out that the omission of a runtime
>exception is effectively a mistake, not some sort of odd trick to get

Right, I wouldn’t have guessed otherwise, I was just surprised.

>access to your source. I realize that doesn't make strict licensing
>possible, but it makes most people feel a bit better about using Open
>Watcom

The problem here is that the source is under another copyleft licence
so while you already have access to it, I cannot distribute binaries
compiled with Open Watcom of it *at all* — this applies to all strong
copyleft licences like the GPL but possibly others.

Thanks anyway.

**Comment by ghost** [comment]
I have always taken it as 'no news is good news'.  As in, the omission of a runtime exception means that there is no restriction upon the binaries generated.  Generally, it can be interpreted as this.

Otherwise, it would be a completely unusable suite.

**Comment by mirabilos** [comment]
Adam dixit:

>I have always taken it as 'no news is good news'.  As in, the omission
>of a runtime exception means that there is no restriction upon the
>binaries generated. Generally, it can be interpreted as this.

Unfortunately, this is not correct. Any creator of a work automatically
has rights in it. As long as part of the work is reproduced in the
generated binaries (which, while compiler output isn’t, the startup
files and run-time library are) the licence needs to explicitly state
how such impacts the executables created, the default being that the
conditions of the licence on the run-time library apply.

>Otherwise, it would be a completely unusable suite.

That’s my point.

**Comment by ghost** [comment]
I know it is not correct in a strict sense.  I know about licence compatibility etc..  but the point I was trying to make, is that there is nothing you can do if you want an absolute assurance you're ok.  Well, you're not going to get it.  But many people use Open Watcom for things, I believe even Retro City Rampage used it for the DOS binaries and nobody is chasing them down asking for the source etc..

Nobody will bother you, the original company is gone, the new company doesn't care, so enjoy it or use DJGPP or something.

**Comment by mirabilos** [comment]
The original author of the GPL’d software I want to patch and make available can bother me, and DJGPP is for 32-bit ☹

**Comment by ghost** [comment]
Isn't DOS a 16 and 32bit OS anyway?  I've never heard of a 64bit DOS..

**Comment by sezero** [comment]
He means that DJGPP can't build a 16 bit program.

**Comment by ghost** [comment]
Aha, of course (since he mentioned Turbo C earlier).  Seems you're almost out of options..

**Comment by tbullock** [comment]
Reading the license for distributed binaries it seems that the options here are as follows:

1. Rewrite the runtime libraries with more permissive license terms
2. Reach out to SAP management/legal for a v2 license

**Comment by ArmstrongJ** [comment]
Well 1 just isn't going to happen (it'd be a laughably large task).  So someone should feel free to contact whoever the hell Sybase is at this point (I guess it's SAP...) and ask them to declare that the Open Watcom Public License v 2 is just a 3-clause BSD or something.  If anyone is waiting for someone else to do that, though, **don't.**  Whoever wants the license changed should contact Sybase/whoever.  Nobody else is going to do it.

**Comment by ssokolow** [comment]
For the record, I don't know whether it'll meet your needs but, if you haven't tried it, [gcc-ia16](https://github.com/tkchia/gcc-ia16) exists as a GCC-based compiler capable of building real-mode DOS executables.
